File: DotNetGotchas/VBDN/ES/Modified1/ESPlatformDiff/ESLib/Factory.vb
' Factory.vb part of ESLib.dll
Imports System.EnterpriseServices

Public Interface ITransactionCoordinator
    Sub SetVoteToAbort()
End Interface

<Transaction(TransactionOption.Required), JustInTimeActivation()> _
Public Class Factory
    Inherits ServicedComponent
    Implements ITransactionCoordinator

    Public Function CreateComp(ByVal key As Integer) As Comp
        ContextUtil.MyTransactionVote = TransactionVote.Abort

        Dim theComp As New Comp

        theComp.TheTransactionCoordinator = Me

        theCOmp.init(key)

        ContextUtil.MyTransactionVote = TransactionVote.Commit

        Return theCOmp
    End Function
    Protected Overloads Overrides Sub Dispose( _
        ByVal disposing As Boolean)
        If disposing Then
            ContextUtil.DeactivateOnReturn = True
        End If

        MyBase.Dispose(disposing)
    End Sub

    Public Sub SetVoteToAbort() _
        Implements ITransactionCoordinator.SetVoteToAbort
        ContextUtil.MyTransactionVote() = TransactionVote.Abort
    End Sub
End Class

File: DotNetGotchas/CSharp/ES/Modified1/ESPlatformDiff/ESLib/Factory.cs
// Factory.cs part of ESLib.dll
using System;
using System.EnterpriseServices;

namespace ESLib
{
	public interface ITransactionCoordinator
	{
		void SetVoteToAbort();
	}

	[Transaction(TransactionOption.Required), JustInTimeActivation]
	public class Factory : ServicedComponent, ITransactionCoordinator
	{
		public Comp CreateComp(int key)
		{
			ContextUtil.MyTransactionVote 
				= TransactionVote.Abort;

			Comp theComp = new Comp();
			
			theComp.TheTransactionCoordinator = this;

			theComp.init(key);

			ContextUtil.MyTransactionVote 
				= TransactionVote.Commit;

			return theComp;
		}

		protected override void Dispose(bool disposing)
		{
			if (disposing)
			{
				ContextUtil.DeactivateOnReturn = true;
			}
			base.Dispose (disposing);
		}

		#region ITransactionCoordinator Members

		public void SetVoteToAbort()
		{
			ContextUtil.MyTransactionVote 
				= TransactionVote.Abort;
		}

		#endregion
	}
}
